keypair/src/lib.rs

```diff
@@ -101,35 +98,28 @@ impl Keypair {
     /// Only use this in tests or when strictly required. Consider using [`std::sync::Arc<Keypair>`]
     /// instead.
     pub fn insecure_clone(&self) -> Self {
-        Self(ed25519_dalek::Keypair {
-            // This will never error since self is a valid keypair
-            secret: ed25519_dalek::SecretKey::from_bytes(self.0.secret.as_bytes()).unwrap(),
-            public: self.0.public,
-        })
+        Self(self.0.clone())
     }
 }
 
 impl TryFrom<&[u8]> for Keypair {
     type Error = SignatureError;
 
     fn try_from(bytes: &[u8]) -> Result<Self, Self::Error> {
-        if bytes.len() < ed25519_dalek::KEYPAIR_LENGTH {
+        let Some(keypair_slice) = bytes.get(..ed25519_dalek::KEYPAIR_LENGTH) else {
             return Err(SignatureError::from_source(String::from(
                 "candidate keypair byte array is too short",
             )));
-        }
-        let secret =
-            ed25519_dalek::SecretKey::from_bytes(&bytes[..ed25519_dalek::SECRET_KEY_LENGTH])
-                .map_err(SignatureError::from_source)?;
-        let public =
-            ed25519_dalek::PublicKey::from_bytes(&bytes[ed25519_dalek::SECRET_KEY_LENGTH..])
-                .map_err(SignatureError::from_source)?;
-        let expected_public = ed25519_dalek::PublicKey::from(&secret);
-        (public == expected_public)
-            .then_some(Self(ed25519_dalek::Keypair { secret, public }))
-            .ok_or(SignatureError::from_source(String::from(
-                "keypair bytes do not specify same pubkey as derived from their secret key",
-            )))
+        };
+        // compiler elides the second len check so try_into().unwrap() is as fast as unsafe code
+        let keypair_bytes: &[u8; ed25519_dalek::KEYPAIR_LENGTH] = keypair_slice.try_into().unwrap();
```

Comment:
What do you think about merging the length check with this line? It seems like the code will be slightly more concise and we can remove the unwrap as well.

```
let keypair_bytes: &[u8; ed25519_dalek::KEYPAIR_LENGTH] = bytes
    .try_into()
    .map_err(|_| SignatureError::from_source(String::from(
        "candidate keypair byte array is the wrong length",
    )))?;
```

Reply:
Yeah if we can allow breaking change, then I would vote for a more strict behavior for safety. I don't see a clear use case for a keypair that is represented by more than 64 bytes. The current implementation silently truncates the input, which can hide bugs.